File: src/main/java/one/pkg/modpublish/util/protect/Protect.java
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package one.pkg.modpublish.util.protect;

import com.intellij.ui.components.JBTextField;
import org.jetbrains.annotations.NotNull;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.security.spec.KeySpec;
import java.util.Base64;

public class Protect {
    private static final int GCM_IV_LENGTH = 12;
    private static final int GCM_TAG_LENGTH = 16;

    private static SecretKey deriveKey(String hash) throws Exception {
        byte[] salt = new byte[16];
        byte[] hashBytes = hash.getBytes(StandardCharsets.UTF_8);
        System.arraycopy(hashBytes, 0, salt, 0, Math.min(hashBytes.length, 16));

        KeySpec spec = new PBEKeySpec(hash.toCharArray(), salt, 100000, 256);
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        byte[] key = factory.generateSecret(spec).getEncoded();
        return new SecretKeySpec(key, "AES");
    }

    public static String encryptString(String data, String hash) {
        try {
            SecretKey key = deriveKey(hash);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");

            byte[] iv = new byte[GCM_IV_LENGTH];
            new SecureRandom().nextBytes(iv);
            GCMParameterSpec parameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv);

            cipher.init(Cipher.ENCRYPT_MODE, key, parameterSpec);
            byte[] encryptedBytes = cipher.doFinal(data.getBytes(StandardCharsets.UTF_8));

            byte[] encryptedWithIv = new byte[GCM_IV_LENGTH + encryptedBytes.length];
            System.arraycopy(iv, 0, encryptedWithIv, 0, GCM_IV_LENGTH);
            System.arraycopy(encryptedBytes, 0, encryptedWithIv, GCM_IV_LENGTH, encryptedBytes.length);

            return Base64.getEncoder().encodeToString(encryptedWithIv);
        } catch (Exception e) {
            return data;
        }
    }

    public static @NotNull String decryptString(JBTextField text) {
        return decryptString(text.getText());
    }

    public static @NotNull String decryptString(String encryptedData) {
        return decryptString(encryptedData, HardwareFingerprint.generateSecureProjectKey());
    }

    public static @NotNull String decryptString(String encryptedData, String hash) {
        try {
            SecretKey key = deriveKey(hash);
            byte[] encryptedWithIv = Base64.getDecoder().decode(encryptedData);

            byte[] iv = new byte[GCM_IV_LENGTH];
            System.arraycopy(encryptedWithIv, 0, iv, 0, GCM_IV_LENGTH);

            byte[] encrypted = new byte[encryptedWithIv.length - GCM_IV_LENGTH];
            System.arraycopy(encryptedWithIv, GCM_IV_LENGTH, encrypted, 0, encrypted.length);

            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            GCMParameterSpec parameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv);
            cipher.init(Cipher.DECRYPT_MODE, key, parameterSpec);

            byte[] decryptedBytes = cipher.doFinal(encrypted);
            return new String(decryptedBytes, StandardCharsets.UTF_8);
        } catch (Exception e) {
            return encryptedData;
        }
    }
}

File: src/main/kotlin/one/pkg/modpublish/util/protect/Protect.kt
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package one.pkg.modpublish.util.protect

import com.intellij.ui.components.JBTextField
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.security.spec.KeySpec
import java.util.*
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

@Suppress("UNUSED")
object Protect {
    private const val GCM_IV_LENGTH = 12
    private const val GCM_TAG_LENGTH = 16

    private fun deriveKey(hash: String): SecretKey {
        val salt = ByteArray(16).apply {
            val hashBytes = hash.toByteArray(StandardCharsets.UTF_8)
            System.arraycopy(hashBytes, 0, this, 0, hashBytes.size.coerceAtMost(16))
        }
        val spec: KeySpec = PBEKeySpec(hash.toCharArray(), salt, 100_000, 256)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val key = factory.generateSecret(spec).encoded
        return SecretKeySpec(key, "AES")
    }

    fun encryptString(data: String, hash: String): String = runCatching {
        val key = deriveKey(hash)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val iv = ByteArray(GCM_IV_LENGTH).apply { SecureRandom().nextBytes(this) }
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(GCM_TAG_LENGTH * 8, iv))
        val encryptedBytes = cipher.doFinal(data.toByteArray(StandardCharsets.UTF_8))
        (iv + encryptedBytes).let { Base64.getEncoder().encodeToString(it) }
    }.getOrDefault(data)

    fun JBTextField.decryptString(): String = decryptString(this.text)

    fun decryptString(encryptedData: String): String =
        decryptString(encryptedData, HardwareFingerprint.secureProjectKey)

    fun decryptString(encryptedData: String, hash: String): String = runCatching {
        val key = deriveKey(hash)
        val encryptedWithIv = Base64.getDecoder().decode(encryptedData)
        val iv = encryptedWithIv.copyOfRange(0, GCM_IV_LENGTH)
        val encrypted = encryptedWithIv.copyOfRange(GCM_IV_LENGTH, encryptedWithIv.size)
        Cipher.getInstance("AES/GCM/NoPadding").apply {
            init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(GCM_TAG_LENGTH * 8, iv))
        }.doFinal(encrypted).toString(StandardCharsets.UTF_8)
    }.getOrDefault(encryptedData)
}
